src/main/java/com/google/finapp/SpannerDao.java

```diff
@@ -16,133 +16,183 @@
 
 import com.google.cloud.ByteArray;
 import com.google.cloud.spanner.*;
-import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.inject.Inject;
 
 import java.math.BigDecimal;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.Arrays;
 
 final class SpannerDao {
 
-  private final DatabaseClient databaseClient;
+  private final String projectId;
+  private final String databaseId;
+  private final String instanceId;
+  private final String connectionUrl;
 
   @Inject
-  SpannerDao(DatabaseClient databaseClient) {
-    this.databaseClient = databaseClient;
+  SpannerDao(@ArgsModule.SpannerProjectId String spannerProjectId,
+      @ArgsModule.SpannerInstanceId String spannerInstanceId,
+      @ArgsModule.SpannerDatabaseId String spannerDatabaseId) {
+    this.projectId = spannerProjectId;
+    this.databaseId = spannerDatabaseId;
+    this.instanceId = spannerInstanceId;
+    // use this URL to connect to the emulator
+    this.connectionUrl = String.format(
+        "jdbc:cloudspanner://localhost:9010/projects/%s/instances/%s/databases/%s;usePlainText=true",
+        projectId, instanceId, databaseId);
+    // use this URL to connect to Cloud Spanner
+    // this.connectionUrl = String.format(
+    //     "jdbc:cloudspanner:/projects/%s/instances/%s/databases/%s",
+    //     projectId, instanceId, databaseId);
   }
 
-  void createCustomer(ByteArray customerId, String name, String address) throws SpannerException {
-    databaseClient.write(
-        ImmutableList.of(
-            Mutation.newInsertBuilder("Customer")
-                .set("CustomerId")
-                .to(customerId)
-                .set("Name")
-                .to(name)
-                .set("Address")
-                .to(address)
-                .build()));
+  void createCustomer(ByteArray customerId, String name, String address) throws SQLException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+      try (PreparedStatement ps =
+          connection.prepareStatement(
+              "INSERT INTO Customer\n"
+                  + "(CustomerId, Name, Address)\n"
+                  + "VALUES\n"
+                  + "(?, ?, ?)")) {
+        ps.setBytes(1, customerId.toByteArray());
+        ps.setString(2, name);
+        ps.setString(3, address);
+        ps.executeUpdate();
+        System.out.printf("Customer created: %s\n", name);
```

Comment:
Eventually we should use a java.util.logging.Logger for such information. Something like:

1. Create the logger in the class constructor:
private static final Logger logger = Logger.getLogger(TheClassName.class.getName());

2. Here you can do something like:
logger.log(Level.INFO, "My amazing message");

Reply:
Yes, this is a good idea. This should be handled in a future PR.

Reply:
https://github.com/GoogleCloudPlatform/cloud-spanner-samples/issues/11